### PR TITLE
API docs: link resultSet -> documentationResult

### PR DESCRIPTION
**Pull request by slimsag**
This PR links from the existing `resultSet` for a range to the corresponding `documentationResult`, for var/const/func/type definitions, using a `documentationResult` edge.

There are a number of use cases for this data:

1. Wherever you'd normally get just a `Hover`, `Go to definition` or `Find references` option - you could also get an `API docs` option which jumps you to the API docs for the corresponding symbol. I'd like to add this in the not too distant future.
2. By looking up the reversed link (documentationResult -> resultSet) you can find the corresponding hover/definition/references for a symbol being described in API docs which has a few benefits:
    a. _references_ make a great set of _usage examples_ to show for a symbol in API docs.
    b. _definition_ lets one jump from API docs -> the actual code (and vice-versa).

---

Consideration: which way should the edge be drawn? We need both directions of data.

* A single `resultSet -> documentationResult` edge (what I chose here.)
* A single `documentationResult -> resultSet` edge.
* Both

All three options give us the same data (_we do need both directions of data_), so I just went with the easier to implement option.

### CODE REVIEW

**Comment by slimsag** [comment]
Note: see the first commit, the **vast** majority of lines changed/added here are actually just to make testing this easy/nice. I've chosen to emit the hover/reference/definition info as part of the markdown as a sort of snapshot testing approach - which seems quite nice and lets you easily explore/validate the data. Here's an example: https://gist.github.com/slimsag/60e1a39b601d2262df289806ba8b148d

File: internal/indexer/doctomarkdown.go [hunk 1]
```diff
@@ -34,12 +36,25 @@ func doctomarkdown(ctx context.Context, r io.Reader, matchingTags []protocol.Doc
 	return buf.String(), nil
 }
 
+type DocsRefDefInfo struct {
+	Document string
+	Ranges   []reader.Range
+}
+
 type converter struct {
```
Comment (efritz) [inline internal/indexer/doctomarkdown.go:44]: Since this whole file is used only in tests, can we give it a `_test.go` suffix?

**Comment by efritz** [comment]
@slimsag It doesn't matter which direction the edges point once we've correlated the graph. We can easily invert a relationship in the backend if we need to query the reverse direction since we slurp the entire index during processing anyway.

**Review by efritz — APPROVED** [review]
🎉

File: internal/indexer/documentation.go [hunk 2]
```diff
@@ -32,20 +33,37 @@ import (
...
+		d                     = &docsIndexer{i: i}
+		mu                    sync.Mutex
+		docsPackages          []docsPackage
+		emitted               = make(emittedDocumentationResults, 4096)
+		emittedPackagesByPath = make(map[string]uint64, 32)
+		errs                  error
 	)
 	i.visitEachPackage("Indexing documentation", func(p *packages.Package) {
 		docsPkg, err := d.indexPackage(p)
+		mu.Lock()
```
[the remark this answers is not in the document]
Comment (slimsag) [reply internal/indexer/documentation.go:58]: `visitEachPackage` is called in parallel, and the multierror `errors.Append` below would be racy if not protected by a mutex as well.

[the remark this answers is not in the document]
Comment (efritz) [reply internal/indexer/documentation.go:58]: Can you put this explanation in source? 😛

**Pull request merged**